# Part 1


## Explain the variables contained in the dataset. What do they refer to?

1. **ID**: An ID variable is a variable that assigns a unique value to each entity in a dataset. ID variables should to be unique. 
2. **Date**: Based on the dataset description from Kaggle, it can be understood that the dataset contains house sale prices for King County, including Seattle. It includes homes sold between May 2014 and May 2015. Therefore, the dataset refers to sold houses within that period.
3. **Price**: The selling price of the house.
4. **Bedrooms**: The number of bedrooms in the house.
5. **Bathrooms**: The number of bathrooms in the house.
6. **sqft_living**: The square footage of the living area.
7. **sqft_lot**: The lot's square footage, which is the area on which a home or other building is built.
8. **Floors**: The number of floors in the house.
9. **Waterfront**: Whether or not the house has a waterfront.
10. **View**: Whether the view from the house is good or bad.
11. **Condition**: The overall condition of the house.
12. **Grade**: Represent the overall quality of the house.
13. **sqft_above**: The square footage of the living area above ground level.
14. **sqft_basement**: The square footage of the basement.
15. **yr_built**: The year the house was built.
16. **yr_renovated**: The year the house wase renovated.
17. **zipcode**: The zipcode of the house.
18. **lat**: The latitude of the house. 
19. **long**: The longitude of the house.
20. **sqft_living15**: The house's living area's square footage in 2015. Understanding how the size of the house has changed over time can be enabled by this information.
21. **sqft_lot15**: The square footage of the lot that the house is located on in 2015.

## What are their specificities? Explain their distribution, most common values, as well as the largest, and smallest values.

In [2]:
import pandas as pd

In [67]:
df = pd.read_csv('kc_house_data.csv')

In [4]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

In [5]:
df.shape[0]

21613

**We have 21613 house in this dataset.**

In [6]:
#Most common number of bedrooms in the dataset
df['bedrooms'].value_counts().sort_values(ascending=False).index[0]

3

In [7]:
#Most common number of bathrooms
df['bathrooms'].value_counts().sort_values(ascending=False).index[0]

2.5

In [8]:
#Average price of a house 
df['price'].mean()

540182.1587933188

In [9]:
#The house with the highest price
df['price'].max()

7700000.0

In [10]:
#The house with the lowest price
df['price'].min()

75000.0

In [12]:
#The house with the most square footage of living area (The Largest footage of living area)
df['sqft_living'].max()

13540

In [16]:
#Price of the house with the most square footage of living area
df.nlargest(1, 'sqft_living')['price']

12777    2280000.0
Name: price, dtype: float64

In [24]:
df1 = df[df.sqft_living ==13540]

In [25]:
df1['price']

12777    2280000.0
Name: price, dtype: float64

In [31]:
#Number of bedrooms of the house with the most square footage of living area
df.nlargest(1, 'sqft_living')['bedrooms']

12777    8.0
Name: bathrooms, dtype: float64

In [15]:
#The house with the lowest square footage of living area
df['sqft_living'].min()

290

In [27]:
#Price of the house with the lowest square footage of living area
df.nsmallest(1,'sqft_living')['price']

19452    142000.0
Name: price, dtype: float64

In [30]:
#Number of bedrooms of the house with the lowest square footage of living area
df.nsmallest(1, 'sqft_living')['bedrooms']

19452    0
Name: bedrooms, dtype: int64

**The amount of living space in a home affects how much it costs. Homes with larger living spaces typically cost more, whereas homes with smaller living spaces typically cost less. This is due to the fact that larger homes frequently have more rooms. On the other hand, families or individuals seeking more compact living arrangements may find smaller homes to be more affordable options.**

In [35]:
#The view of the house with the most square footage of living area
df.nlargest(1, 'sqft_living')['view']

12777    4
Name: view, dtype: int64

In [34]:
#The view of the house with the lowes quare footage of living area
df.nsmallest(1, 'sqft_living')['view']

19452    0
Name: view, dtype: int64

**Houses with large living areas and views command higher prices. Additionally, larger houses are more likely to have desirable views, as they are often built on more prominent plots of land.**

In [37]:
#Get the number of houses whether with or without waterfront
df['waterfront'].value_counts()

0    21450
1      163
Name: waterfront, dtype: int64

In [38]:
#Average price of houses with water front
df[df['waterfront'] == 1]['price'].mean()

1662524.1840490797

In [39]:
#Average price of houses without water front
df[df['waterfront'] == 0]['price'].mean()

531653.4058741259

- **Houses with waterfront are more rare than houses without waterfront.**
- **Houses with waterfront are more expensive than houses without waterfront.**

In [40]:
#Get the value of condition
df['condition'].value_counts()

3    14031
4     5679
5     1701
2      172
1       30
Name: condition, dtype: int64

- **The condition column can have 5 values: 1, 2, 3, 4, or 5.**
- **The most common condition among the sold houses is condition 3, with 14,031 occurrences. This indicates that a significant number of houses are in relatively average condition.**

In [42]:
#Average price of houses with good conditions
df[df['condition']==4]['price'].mean()

521300.70522979397

In [44]:
#Average price of house with poor conditions
df[df['condition']==1]['price'].mean()

334431.6666666667

- **Houses in good condition are more expensive than houses in poor condition.**
- **The condition of the house have an impact on its price.**

In [46]:
#Get the counts of floors
df['floors'].value_counts()

1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: floors, dtype: int64

**The most common type of houses in the dataset are single-floor houses.**

In [54]:
#Average price of single-floor
df[df['floors']==1]['price'].mean()

442219.6138576779

In [53]:
#Average price three or more floors
df[df['floors']>3]['price'].mean()

933937.5

**Houses with three or more floors are often more expensive**

## Try to look for geographical or temporal patterns in the time, latitude, and longitude variables.

In [55]:
#Get the oldest house
df['yr_built'].min()

1900

In [56]:
#Get the newest house
df['yr_built'].max()

2015

In [58]:
#Get the house with the most recent renovation
df['yr_renovated'].max()

2015

In [59]:
#Get the oldest renovation
df['yr_renovated'].min()

0

In [60]:
df['yr_renovated'].value_counts()

0       20699
2014       91
2013       37
2003       36
2005       35
        ...  
1951        1
1959        1
1948        1
1954        1
1944        1
Name: yr_renovated, Length: 70, dtype: int64

- **Most houses were not renovated.**
- **Houses built in the 2000s were more likely to be renovated.**
- **Houses built in the 1900 were less likely to be renovated.**

In [62]:
# Get the house with the highest latitude
df['lat'].max()

47.7776

In [66]:
#Get the house with the lowest latitude
df['lat'].min()

47.1559

In [64]:
#Get the house with the highest longitude
df['long'].max()

-121.315

In [65]:
#Get the house with the lowest longitude
df['long'].min()

-122.519

**The houses are located in a relatively small area**

# Part 2

##  Read documentation: https://pandas.pydata.org/docs/index.html

## Go to the parts we have covered today and study all the details about each function used, write down examples with explanations

- **DataFrame.info(verbose=None, buf=None, max_cols=None, memory_usage=None, show_counts=None)**

In [68]:
# Verbos (bool) (whether to print the full summary)
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [71]:
#If verbose == false it just print out a summary
df.info(verbose=False)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Columns: 21 entries, id to sqft_lot15
dtypes: float64(5), int64(15), object(1)
memory usage: 3.5+ MB


In [72]:
#Without deep introspection, memory usage is estimated based on the column data types and the number of rows. With deep introspection, the actual memory usage is calculated, which can be more accurate
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

- **DataFrame.value_counts(subset=None, normalize=False, sort=True, ascending=False, dropna=True)**

In [74]:
#Soting in ascending order based on the number of occurrence
df['bedrooms'].value_counts(ascending=True)

11       1
33       1
10       3
9        6
0       13
8       13
7       38
1      199
6      272
5     1601
2     2760
4     6882
3     9824
Name: bedrooms, dtype: int64

In [75]:
#Return proportions rather than frequencies.
df['bedrooms'].value_counts(normalize=True)

3     0.454541
4     0.318419
2     0.127701
5     0.074076
6     0.012585
1     0.009207
7     0.001758
0     0.000601
8     0.000601
9     0.000278
10    0.000139
11    0.000046
33    0.000046
Name: bedrooms, dtype: float64

- **DataFrame.mean(axis=0, skipna=True, numeric_only=False, **kwargs)****

In [82]:
# Calculate the mean of the price column, ignoring NaN values
df['price'].mean(axis=0, skipna=True)

540182.1587933188

- **DataFrame.min(axis=0, skipna=True, numeric_only=False, **kwargs)****

In [84]:
# Calculate the minimum of the sqft_living column, ignoring NaN values
df['sqft_living'].min(axis=0, skipna=True)

290

- **DataFrame.nsmallest(n, columns, keep='first')**

keep{‘first’, ‘last’, ‘all’}, default ‘first’
Where there are duplicate values:

first : take the first occurrence.

last : take the last occurrence.

all : do not drop any duplicates, even it means selecting more than n items.
**Returns**: DataFrame
             The first n rows ordered by the given columns in descending order.
*`(Same for nlargest)`*

In [85]:
df.nsmallest(3, 'bedrooms', keep='all')

id             date      price  bedrooms  bathrooms  \
875    6306400140  20140612T000000  1100000.0         0       0.00   
3119   3918400017  20150205T000000   380000.0         0       0.00   
3467   1453602309  20140805T000000   288000.0         0       1.50   
4868   6896300380  20141002T000000   228000.0         0       1.00   
6994   2954400190  20140624T000000  1300000.0         0       0.00   
8477   2569500210  20141117T000000   339950.0         0       2.50   
8484   2310060040  20140925T000000   240000.0         0       2.50   
9773   3374500520  20150429T000000   355000.0         0       0.00   
9854   7849202190  20141223T000000   235000.0         0       0.00   
12653  7849202299  20150218T000000   320000.0         0       2.50   
14423  9543000205  20150413T000000   139950.0         0       0.00   
18379  1222029077  20141029T000000   265000.0         0       0.75   
19452  3980300371  20140926T000000   142000.0         0       0.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
875           3064      4764     3.5           0     2  ...      7   
3119          1470       979     3.0           0     2  ...      8   
3467          1430      1650     3.0           0     0  ...      7   
4868           390      5900     1.0           0     0  ...      4   
6994          4810     28008     2.0           0     0  ...     12   
8477          2290      8319     2.0           0     0  ...      8   
8484          1810      5669     2.0           0     0  ...      7   
9773          2460      8049     2.0           0     0  ...      8   
9854          1470      4800     2.0           0     0  ...      7   
12653         1490      7111     2.0           0     0  ...      7   
14423          844      4269     1.0           0     0  ...      7   
18379          384    213444     1.0           0     0  ...      4   
19452          290     20875     1.0           0     0  ...      1   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
875          3064              0      1990             0    98102  47.6362   
3119         1470              0      2006             0    98133  47.7145   
3467         1430              0      1999             0    98125  47.7222   
4868          390              0      1953             0    98118  47.5260   
6994         4810              0      1990             0    98053  47.6642   
8477         2290              0      1985             0    98042  47.3473   
8484         1810              0      2003             0    98038  47.3493   
9773         2460              0      1990             0    98031  47.4095   
9854         1470              0      1996             0    98065  47.5265   
12653        1490              0      1999             0    98065  47.5261   
14423         844              0      1913             0    98001  47.2781   
18379         384              0      2003             0    98070  47.4177   
19452         290              0      1963             0    98024  47.5308   

          long  sqft_living15  sqft_lot15  
875   -122.322           2360        4000  
3119  -122.356           1470        1399  
3467  -122.290           1430        1650  
4868  -122.261           2170        6000  
6994  -122.069           4740       35061  
8477  -122.151           2500        8751  
8484  -122.053           1810        5685  
9773  -122.168           2520        8050  
9854  -121.828           1060        7200  
12653 -121.826           1500        4675  
14423 -122.250           1380        9600  
18379 -122.491           1920      224341  
19452 -121.888           1620       22850  

[13 rows x 21 columns]

In [86]:
df.nsmallest(3, 'bedrooms', keep='first')

id             date      price  bedrooms  bathrooms  \
875   6306400140  20140612T000000  1100000.0         0        0.0   
3119  3918400017  20150205T000000   380000.0         0        0.0   
3467  1453602309  20140805T000000   288000.0         0        1.5   

      sqft_living  sqft_lot  floors  waterfront  view  ...  grade  sqft_above  \
875          3064      4764     3.5           0     2  ...      7        3064   
3119         1470       979     3.0           0     2  ...      8        1470   
3467         1430      1650     3.0           0     0  ...      7        1430   

      sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
875               0      1990             0    98102  47.6362 -122.322   
3119              0      2006             0    98133  47.7145 -122.356   
3467              0      1999             0    98125  47.7222 -122.290   

      sqft_living15  sqft_lot15  
875            2360        4000  
3119           1470        1399  
3467           1430        1650  

[3 rows x 21 columns]

In [87]:
df.nsmallest(3, 'bedrooms', keep='last')

id             date     price  bedrooms  bathrooms  \
19452  3980300371  20140926T000000  142000.0         0       0.00   
18379  1222029077  20141029T000000  265000.0         0       0.75   
14423  9543000205  20150413T000000  139950.0         0       0.00   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
19452          290     20875     1.0           0     0  ...      1   
18379          384    213444     1.0           0     0  ...      4   
14423          844      4269     1.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
19452         290              0      1963             0    98024  47.5308   
18379         384              0      2003             0    98070  47.4177   
14423         844              0      1913             0    98001  47.2781   

          long  sqft_living15  sqft_lot15  
19452 -121.888           1620       22850  
18379 -122.491           1920      224341  
14423 -122.250           1380        9600  

[3 rows x 21 columns]

# Part 3

## Create a random Pandas dataframe from a dictionary. Show several examples

In [95]:
import pandas as pd
import numpy as np

In [100]:
#Get random dictionary
random_dict = {
    'Country': ['USA', 'Canada', 'UK', 'Australia'],
    'Population': np.random.randint(1000000, 100000000, size=4),
    'GDP': np.random.uniform(1000, 10000, size=4)
}

print(data)

{'Country': ['USA', 'Canada', 'UK', 'Australia'], 'Population': array([44200287, 37284599,  9037208, 95801498]), 'GDP': array([3193.85243904, 2383.64563098, 5715.21915418, 9253.97140114])}


In [101]:
# Create a DataFrame from the dictionary
dataframe1 = pd.DataFrame(random_dict)

In [102]:
dataframe1.head()

Country  Population          GDP
0        USA    17065142  8393.686761
1     Canada     8439224  2442.805443
2         UK    65734796  1236.562836
3  Australia    27473600  1115.258506

In [104]:
data = {'col_1': [3, 2, 1, 0], 'col_2': ['a', 'b', 'c', 'd']}
pd.DataFrame(data).head()

col_1 col_2
0      3     a
1      2     b
2      1     c
3      0     d

## Part 4

## Read about data types, specifically in Pandas. Write examples of how to change from one data type to another using the dataframes created above

<img src="Screenshot (1457).png" alt="Image Alt Text" width="1000" Hight="1000">

In order to convert data types in pandas, there are three basic options:

- Use astype() to force an appropriate dtype
- Create a custom function to convert the data
- Use pandas functions such as **to_numeric()** or **to_datetime()**

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [109]:
#Change the data type of the waterfront column into boolean
df['waterfront'] = df['waterfront'].astype('bool')

In [110]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  bool   
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [111]:
df['floors'].unique()

array([1. , 2. , 1.5, 3. , 2.5, 3.5])

In [112]:
df['waterfront'] = df['waterfront'].astype('category')

- **Int64**: Is a data type that represents whole numbers (integers) without any fractional part. It uses 64 bits to store each integer value, which gives it a range of values from -9.2 quintillion to 9.2 quintillion. int64 is suitable for representing discrete values like counts, IDs, and indices.
- **Float64**: Is a data type that represents numbers with a fractional part and can also represent integers. It uses 64 bits to store each floating-point value, which gives it a wider range of values than int64. float64 is suitable for representing continuous values like measurements, quantities, and calculated results.